src/Lorentz/Contracts/BaseDAO/Proposal.hs

```diff
@@ -727,3 +735,93 @@ flush config = do
 
   iter (handleProposalIsOver config)
   swap; pair
+
+deleteProposals
+  :: forall pm s. NiceParameter pm
+  => Natural : Map Timestamp (ProposalKey pm) : s
+  :-> Map Timestamp (ProposalKey pm)  : s
+deleteProposals = do
+  toNamed #deleteAmount
+  swap; dup; 
+  dip $ do
+    swap
+    push 0; toNamed #counter
+
+  stackType @(Map Timestamp (ProposalKey pm) : ("counter" :! Natural) : ("deleteAmount" :! Natural) : Map Timestamp (ProposalKey pm) : s)
+
+  iter $ do
+    dig @2; dig @2;
+    dupTop2
+    if #counter <. #deleteAmount then do
+      stackType @(("counter" :! Natural) : ("deleteAmount" :! Natural) : (Timestamp, (ProposalKey pm)) : Map Timestamp (ProposalKey pm) : s)
+      -- Delete the proposal
+      dig @3;
+      dig @3;
+      car; dip none; update
+      swap; fromNamed #counter; push (1 :: Natural); add; toNamed #counter
+      dip swap
+    else do
+      -- Increase counter, do nothing
+      dip $ dip drop
+      fromNamed #counter; push (1 :: Natural); add; toNamed #counter
+
+  drop; drop
+
+-- | 'dropProposals' accept a nat which is an amount of proposals to be droped
+-- sorted by oldest (submitted/started date) to newest.
+-- It is useful when 'flush' stuck on one or more failing proposals due to some
+-- particular reasons.
+dropProposals
+  :: forall store ce pm s.
+     (NiceParameter pm, StorageC store ce pm, HasFuncContext s store (AllFuncs store ce pm))
+  => Config ce pm -> Entrypoint' Natural store s
+dropProposals config = do
+  doc $ DDescription "TODO."
+
+  -- guards
+  dip $ do
+    ensureNotMigrated
+    authorizeAdmin
+  ensureNot (0 :: Natural)
+
+  dip $ do dup; stToField #sProposalKeyListSortByDate
+  deleteProposals
+  stSetField #sProposalKeyListSortByDate
+
+  nil;pair
+
+
+
+-- Have to be divided like this due to Maybe a is not comparable
+ensureMaybeIsNot :: forall a s.
+  ( NiceParameter a, NiceStorage a
+  , NicePackedValue a, KnownValue a, NiceComparable a
+  ) => Maybe a -> (Maybe a : s) :-> (Maybe a : s)
+ensureMaybeIsNot input =
+  let
+    (someCase, nothingCase) = case input of
+      Nothing -> (do drop; nop, failCustom_ #vOTING_PERIOD_OVER)
+      Just a -> ((do
+          toNamed #val
+          push a; toNamed #input
+          if #input ==. #val then
+            failCustom_ #vOTING_PERIOD_OVER -- TODO: forbidden entrypoint parameter
+          else
+            nop
+        ), nop)
+  in do
+    dup
+    ifSome someCase nothingCase
+
+ensureNot :: forall a s.
+  ( NiceParameter a, NiceStorage a
+  , NicePackedValue a, KnownValue a, NiceComparable a
+  ) => a -> (a : s) :-> (a : s)
+ensureNot input = do
+  dup; toNamed #val
+  push input; toNamed #input
+  if #input ==. #val then
+    failCustom_ #vOTING_PERIOD_OVER -- TODO: forbidden entrypoint parameter
+  else
+    nop
+
```

Comment:
Yep, these checks look like a good improvement to the contract semantics. 

Though a bit sad that for such things we pay with operation size which is a problem now. I suppose we could also offload such checks to the middleware via mentioning in the description something like "passing 0 is semantically no-op but is still expensive, please don't call us with 0".